### PR TITLE
Move libunwind directories to src/native/external

### PR DESCRIPTION
**Pull request by am11**
* `src/coreclr/pal/src/libunwind` -> `src/native/external/libunwind`
* `src/coreclr/pal/src/libunwind_mac` -> `src/native/external/libunwind_extras/mac`
* `src/coreclr/pal/src/libunwind/libunwind-version.txt` -> `src/native/external/libunwind-version.txt` 
* `src/coreclr/pal/src/libunwind/src/oop` -> `src/native/external/libunwind_extras/oop`
  * so extra files that we compile as part of libunwind project are in same place (currently they are in two places: `libunwind_mac` and `oop` directories)
* Extracted cmake configs in `src/native/external/libunwind.cmake` and reverted cmake files in vendor directory to upstream's original state (same structure is used for other external libs).

### CODE REVIEW

**Comment by am11** [comment]
cc @jkotas, @janvorli, this is a continuation of https://github.com/dotnet/runtime/pull/62673. I haven't moved llvm-libunwind (https://github.com/dotnet/runtime/tree/f3e4e768d10dc502b0a7e9c417ae269dcf75dbc7/src/coreclr/nativeaot/libunwind) to `external` yet, as its tree is diverged from llvm-project's `release/9.x` branch (those patches were never upstreamed). Since now everything is in single repo; if it doesn't regress in size and functionality, should we try to switch NativeAOT to also use this (HP) libunwind (which supports more architectures -- like s390x -- compared to llvm libunwind)?

**Comment by jkotas** [comment]
> if it doesn't regress in size and functionality, should we try to switch NativeAOT to also use this (HP) libunwind (which supports more architectures -- like s390x -- compared to llvm libunwind)?

Yes, it would be nice to unify it if possible. IIRC, macOS pushed us to use llvm libunwind. HP libunwind did not work on macOS and the macOS-provided libunwind did not work either. I do not remember the details. Does HP libunwind work on macOS?

**Comment by am11** [comment]
Ah, that's still the case. We are only using remote unwind feature of HP libunwind on macOS (and FreeBSD) in CoreCRL, for out-of-context unwinding. For the rest of functionality, we use OS provided libunwind (which Apple contributed as llvm-libunwind to LLVM project in 2013). We could use same mechanism in NativeAOT since both libunwind libraries have matching headers. I haven't looked at NativeAOT usage of libunwind in depth, but from a quick search, @janvorli implemented `unw_get_save_loc` for memory-locations (HP libunwind provides this functionality for both register and memory locations), which we are using in `nativeaot/Runtime/unix/UnixContext.cpp`.

ps - Mono uses it's own implementation https://github.com/dotnet/runtime/blob/b85a435/src/mono/mono/utils/mono-stack-unwinding.h / https://github.com/dotnet/runtime/blob/b85a435/src/mono/mono/utils/mono-context.h. It might be possible to switch NativeAOT to Mono's implementation and get rid of llvm-libunwind dependency that way.

**Comment by janvorli** [comment]
That's funny, I have completely forgotten that I have added the `unw_get_save_loc` to our local copy of that libunwind. Maybe we could actually merge that upstream (I'd look into implement the `unw_get_save_loc` support for location in registers too to make it fully implemented) and then using the LLVM libunwind as the unified version might be also an option. 
The only issue would be the missing support for the s390x and the longsoonarch64 that is now being enabled by a 3rd party. But maybe adding support for that would not be difficult. The LLVM libunwind code is almost completely arch agnostic.

**Comment by am11** [comment]
These are all the llvm-libunwind patches (based on `releases/9.x` branch), in our NativeAOT tree: https://github.com/llvm/llvm-project/compare/release/9.x...am11:dotnet-runtime-patches. On their `main` branch, RISC-V and MIPS64 arch support was added since 9.x.

**Comment by kant2002** [comment]
I’m not sure if I’m qualified to speak with authority, but LLVM libunwind for ARM a bit undeveloped. Overall code looks like it is half done. I was able only fix gross issue like this https://github.com/llvm/llvm-project/commit/08a5ac36b956edeb989b4a65269a829eac26a5a2 In EHABI handling. For me this code is a bit scary

**Comment by am11** [comment]
Support status related to architectures we are supporting in .NET runtime (and the upcoming ones, like loongarch64, riscv64):

**Mono unwinder** - all archs (except loongarch64? shouldn't be hard to implement since both MIPS and RISC-V are supported)
**HP libunwind** - all archs
**LLVM libunwind** - all archs except s390x and loongarch64 (llvm does not dogfood its libunwind)
**GCC unwinder** - all archs except loongarch64 (PR is up for loongarch64; gcc does dogfood its unwinder)

If mono unwinder (currently supporting: xarch, armarch, s390x, mips, riscv etc.) suffice the stack unwinding requirements of coreclr and nativeaot (i.e. something equivalent to `unw_get_save_loc` is supported), we can move it to `src/native/stack-unwinder` and delete both HP and LLVM libunwinds directories from the repo. IMHO, that would be the best case scenario and we won't need to wait for next release of external libraries for next architecture support (rust, too, has its own stack unwinder implementation like mono).

**Comment by jkotas** [comment]
> Mono unwinder - all archs (except loongarch64? shouldn't be hard to implement since both MIPS and RISC-V are supported)

I have discussed this with the Mono folks in the past. Mono unwinder is incomplete - it only handles the unwind sequences that Mono emits, it is not a general purpose unwinder.

**Comment by am11** [comment]
1. This document has the current status of llvm-libunwind https://github.com/llvm/llvm-project/blob/main/libunwind/docs/index.rst#current-status, which notes:

    > Remote unwinding remains as future work.

    IIUC, in cross DAC implementation of CoreCLR, we rely on remote (or out-of-context) unwinding (for Windows->Linux and macOS->Linux diagnostics). This is in addition to `unw_get_save_loc` shortcomings. Adding support for both of these missing features is a non-trivial work. Therefore, removing HP libunwind dependency is not as feasible at this point.

2. HP libunwind, OOTH, does not support formats beyond DWARF; whereas, LLVM libunwind supports COFF/XCOFF, DWARF and CodeView formats. Adding macOS MachO/XCOFF support is a non-trivial work. Therefore, removing LLVM libunwind from NativeAOT is also not feasible at this point.

3. 
    > Mono unwinder is incomplete - it only handles the unwind sequences that Mono emits, it is not a general purpose unwinder.

    Therefore reliance on Mono unwinder in CoreCLR and NativeAOT (in order to remove both LLVM and HP libunwind dependencies) is also not feasible right now.

I think then this PR is ready for review as-is. We can move llvm-libunwind tree under `src/native/external` separately (ideally after upstreaming the local changes).

**Comment by am11** [comment]
@janvorli, I have merged the main branch and update armv6 and loongarch64 changes. PTAL.
Browser-wasm test failures are unrelated (#65335).

**Comment by am11** [comment]
Resolved merge conflicts, green CI. 🌴

File: src/native/external/libunwind.cmake [hunk 1]
```diff
@@ -1,7 +1,10 @@
-# This is a custom file written for .NET Core's build system
-# It overwrites the one found in upstream
+include_directories(${CMAKE_CURRENT_LIST_DIR}/libunwind/include/tdep)
```
Comment (janvorli) [inline src/native/external/libunwind.cmake:1]: This file is missing the license header.

File: src/native/external/libunwind_extras/CMakeLists.txt [hunk 2]
```diff
@@ -145,13 +130,13 @@ if(CLR_CMAKE_HOST_WIN32)
     add_compile_options(/permissive-)
 
     # include paths
-    include_directories(include/tdep)
-    include_directories(include)
-    include_directories(${CMAKE_CURRENT_BINARY_DIR}/include/tdep)
-    include_directories(${CMAKE_CURRENT_BINARY_DIR}/include)
+    include_directories(${CLR_SRC_NATIVE_DIR}/external/libunwind/include/tdep)
```
Comment (janvorli) [inline src/native/external/libunwind_extras/CMakeLists.txt:133]: This file is missing the licensing header.

File: src/native/external/libunwind_extras/configure.cmake [hunk 3]
```diff
@@ -13,13 +10,13 @@ if(CLR_CMAKE_HOST_WIN32)
     # Fake it until support is added
     check_include_files(stdalign.h HAVE_STDALIGN_H)
     if (NOT HAVE_STDALIGN_H)
-        configure_file(include/win/fakestdalign.h.in ${CMAKE_CURRENT_BINARY_DIR}/include/stdalign.h COPYONLY)
+        configure_file(${CLR_SRC_NATIVE_DIR}/external/libunwind/include/win/fakestdalign.h.in ${CMAKE_CURRENT_BINARY_DIR}/include/stdalign.h COPYONLY)
```
Comment (janvorli) [inline src/native/external/libunwind_extras/configure.cmake:13]: This file is missing the licensing header.

**Review by janvorli — COMMENTED** [review]
LGTM, just few missing licensing headers in files that are now internal to our repo.

File: src/native/external/libunwind.cmake (hunk 1 above)
Comment (am11) [reply src/native/external/libunwind.cmake:1]: Looks like we usually do not have license headers in cmake files. Out of roughly 300 cmake files in the repo, only 30 has the license header:
```zsh
#  grand total: cmake files
% git ls-files ':/*.cmake' ':/*CMakeLists.txt' | wc -l
     304

# number of files with the license headers
% git grep 'The .NET Foundation' ':/*.cmake' ':/*CMakeLists.txt' | wc -l
      30
```
all 30 of them are under `src/native/corehost`.

File: src/native/external/libunwind_extras/configure.cmake (hunk 3 above)
Comment (am11) [reply src/native/external/libunwind_extras/configure.cmake:13]: Same as above

File: src/native/external/libunwind_extras/CMakeLists.txt (hunk 2 above)
Comment (am11) [reply src/native/external/libunwind_extras/CMakeLists.txt:133]: Same as above

Comment (janvorli) [reply src/native/external/libunwind_extras/CMakeLists.txt:133]: Ah, you are right, I have incorrectly thought we have it in these files.

**Review by janvorli — APPROVED** [review]
LGTM, thank you!

**Pull request merged**